### PR TITLE
KAZOO-4143: system dial plans

### PR DESCRIPTION
**Pull request by siplabs**

### CODE REVIEW

File: applications/callflow/src/cf_util.erl [hunk 1]
```diff
@@ -733,6 +741,26 @@ maybe_apply_dialplan([Regex|Regexs], DialPlan, Number) ->
             <<Prefix/binary, Root/binary, Suffix/binary>>
     end.
 
+-spec load_system_dialplans(ne_binaries()) -> wh_json:object().
+load_system_dialplans(Names) ->
+    LowerNames = [wh_util:to_lower_binary(Name) || Name <- Names],
+    case whapps_config:get_all_kvs(<<"dialplans">>) of
+        Plans when is_list(Plans) ->
+            lists:foldl(
+              fun({Key, Val}, Acc) ->
```
Comment (fenollp) [inline applications/callflow/src/cf_util.erl]: Move this `fun` into its own private named function, with a name starting with `fold_`

File: applications/crossbar/src/modules/cb_dialplans.erl [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+%% @public
+%% @doc
+%%
+%% @end
+%%--------------------------------------------------------------------
+-spec validate(cb_context:context()) -> cb_context:context().
+validate(Context) ->
+    Doc = wh_json:from_list(whapps_config:get_all_kvs(<<"dialplans">>)),
+    cb_context:setters(Context, [{fun cb_context:set_resp_data/2, Doc}
+                                 ,{fun cb_context:set_resp_status/2, 'success'}]).
```
Comment (fenollp) [inline applications/crossbar/src/modules/cb_dialplans.erl]: Put `]).` on its own line please ( @jamesaimonetti <3 )

**Comment by siplabs** [comment]
@fenollp 
updated

**Comment by fenollp** [comment]
Thanks @siplabs. I wonder why CI isn't running this, now…
Trying something…

**Comment by fenollp** [comment]
Yayfixedit

**Comment by siplabs** [comment]
@fenollp, I realized that `dilaplans` schema may be improved. I want to try it.

**Comment by siplabs** [comment]
@fenollp updated `dialplans` schema.

**Comment by siplabs** [comment]
ready to commit

File: applications/callflow/src/cf_util.erl [hunk 3]
```diff
@@ -6,6 +6,8 @@
 %%% @contributors
 %%%   Karl Anderson
 %%%   James Aimonetti
+%%%   Sponsored by Conversant Ltd,
```
Comment (lazedo) [inline applications/callflow/src/cf_util.erl]: shouldn't this mention the jira ticket ?

**Comment by lazedo** [comment]
generally i like this feature but i do see some issues with it.
i believe the intention here is to apply dialplans to all accounts, but i see its usefulness somewhat limited specially for systems that have accounts from more than one country.
lets say that a system has accounts for germany, uk, france.
in those countries when people dial 0... , it usually means that they want to apply the country specific dialplan (replace 0 with country indicator) which is hard to apply with this implementation because there's no way to differentiate that.

maybe the implementation should be done like call restrictions, or the account/user/device dialplan should have some indicator/pointer to a system dialpan.

this system dialplan could also be of two types, simple regexe as is now, and also a group/list of other dialplans, in a way that a reference to the group dialplan would apply all the others in the list.

**Comment by k-anderson** [comment]
@lazedo 

I believe the use-case is slightly different @siplabs correct me if I am wrong.  I think they want to create a set of dialplans once then re-use them per account.

As in, they would create a dialplan to correct for say Germany in a common place in the system.  For all German clients they would simply apply this common dialplan.

If that is the use-case then it makes sense to me ;)  Hopefully when we finish the libphonenumber integration all this manipulation will not be necessary!

File: applications/callflow/src/cf_util.erl (hunk 3 above)
Comment (k-anderson) [inline applications/callflow/src/cf_util.erl]: @lazedo preferably but not required ;)

**Comment by lazedo** [comment]
rereading it, i realize that the dialplan of the account contains the links to the system dialplan.
my previous comment makes no sense, and since the scope of this is not grouping dialplans (which can be added later), i'm ok to accept this

**Pull request merged**